controller/scheduler/scheduler.go

```diff
@@ -0,0 +1,1216 @@
+		if err := host.StopJob(job.JobID); err != nil {
+			log.Error("error requesting host to stop job", "err", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func jobConfig(req *JobRequest, hostID string) *host.Job {
+	return utils.JobConfig(req.Job.Formation.ExpandedFormation, req.Type, hostID)
+}
+
+func (s *Scheduler) findBestHost(formation *Formation, typ string) (utils.HostClient, error) {
+	log := fnLogger("app.id", formation.App.ID, "release.id", formation.Release.ID, "job.type", typ)
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return nil, err
+	}
+
+	counts := s.jobs.GetHostJobCounts(formation.key(), typ)
+	var minCount int = math.MaxInt32
+	var hostID string
+	for _, host := range hosts {
+		count, ok := counts[host.ID()]
+		if !ok || count < minCount {
+			minCount = count
+			hostID = host.ID()
+		}
+	}
+	if hostID == "" {
+		return nil, fmt.Errorf("Unable to find a host out of %d host(s)", len(hosts))
+	}
+	log.Info(fmt.Sprintf("using host with least %s jobs", typ), "host.id", hostID)
+	return s.Host(hostID)
+}
+
+func (s *Scheduler) getHosts() ([]utils.HostClient, error) {
+	hosts, err := s.Hosts()
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that we're only following hosts that we can discover
+	knownHosts := make(map[string]struct{})
+	for id, hostStream := range s.hostStreams {
+		if hostStream.Err() == nil {
+			knownHosts[id] = struct{}{}
+		} else {
+			// TODO stop unfollowing hosts here once host syncing is built
+			s.unfollowHost(id)
+		}
+	}
+	for _, h := range hosts {
+		hostID := h.ID()
+		delete(knownHosts, hostID)
+		if _, ok := s.hostStreams[hostID]; !ok {
+			s.followHost(h)
+		}
+	}
+	for id := range knownHosts {
+		// TODO stop unfollowing hosts here once host syncing is built
+		s.unfollowHost(id)
+	}
+	if len(hosts) == 0 {
+		log := fnLogger()
+		e := "no hosts found"
+		log.Error(e)
+		return nil, errors.New(e)
+	}
+
+	return hosts, nil
+}
+
+func (s *Scheduler) Stop() error {
+	log := fnLogger()
+	log.Info("stopping scheduler loop")
+	s.stopOnce.Do(func() {
+		close(s.stop)
+	})
+	return nil
+}
+
+func (s *Scheduler) Subscribe(events chan Event) stream.Stream {
+	log := fnLogger()
+	log.Info("adding event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	s.listeners[events] = struct{}{}
+	return &Stream{s, events}
+}
+
+func (s *Scheduler) Unsubscribe(events chan Event) {
+	log := fnLogger()
+	log.Info("removing event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	delete(s.listeners, events)
+}
+
+func (s *Scheduler) SaveJob(job *Job, appName string, status host.JobStatus, metadata map[string]string) bool {
+	switch status {
+	case host.StatusStarting:
+		s.handleJobEvent(job, JobStateStarting)
+	case host.StatusRunning:
+		s.handleJobEvent(job, JobStateRunning)
+	default:
+		if job.IsStopped() {
+			s.handleJobEvent(job, JobStateStopped)
+		} else {
+			s.handleJobCrash(job)
```

Comment:
We need to check whether the job should actually be running before we treat it as crashed and restart it (i.e. by checking the formation), because another scheduler may have stopped the job just before we became leader.